pkg/graveler/ref/manager.go

```diff
@@ -412,19 +445,72 @@ func (m *Manager) ListCommits(ctx context.Context, repositoryID graveler.Reposit
 }
 
 func (m *Manager) FillGenerations(ctx context.Context, repositoryID graveler.RepositoryID) error {
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		return tx.Exec(`WITH RECURSIVE cte AS
-								(
-									SELECT id, 1 AS generation,creation_date
-									FROM graveler_commits
-									WHERE repository_id = $1 AND parents IS NULL OR array_length(parents,1) = 0
-									UNION ALL
-									SELECT DISTINCT(c.id), cte.generation+1 AS generation,c.creation_date
-									FROM graveler_commits c INNER JOIN cte
-									ON cte.id = ANY(c.parents) WHERE repository_id = $1
-								)
-								UPDATE graveler_commits u SET generation = cte.generation FROM cte WHERE u.id = cte.id;`,
-			repositoryID)
+	iter, err := m.ListCommits(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	defer iter.Close()
+
+	commitIDToNode := make(map[graveler.CommitID]CommitNode)
+	var rootCommitID graveler.CommitID
+
+	// make a commits "tree" from map data structure
+	for iter.Next() {
+		commit := iter.Value()
+		if len(commit.Parents) == 0 {
+			rootCommitID = commit.CommitID
+		}
+
+		// adding current node as a child to all parents nodes
+		for _, commitParentID := range commit.Parents {
+			commitParentNode, exist := commitIDToNode[commitParentID]
+			var parentCommitChildrenCommitsIDs = []graveler.CommitID{commit.CommitID}
+			var parentCommitNotVisitedParents int
+			if exist {
+				parentCommitChildrenCommitsIDs = commitParentNode.childCommitsIDs
+				parentCommitChildrenCommitsIDs = append(parentCommitChildrenCommitsIDs, commit.CommitID)
+				parentCommitNotVisitedParents = commitParentNode.notVisitedParents
+			}
+			commitIDToNode[commitParentID] = CommitNode{parentCommitChildrenCommitsIDs, parentCommitNotVisitedParents, 0}
+		}
+
+		currentNode, exist := commitIDToNode[commit.CommitID]
+		var currentCommitChildrenIDs []graveler.CommitID
+		if exist {
+			currentCommitChildrenIDs = currentNode.childCommitsIDs
+		}
+		commitIDToNode[commit.CommitID] = CommitNode{currentCommitChildrenIDs, len(commit.Parents), 0}
+	}
+	if err := iter.Err(); err != nil {
+		return fmt.Errorf("getting value from iterator: %w", err)
+	}
+
+	nodesQueue := make([]graveler.CommitID, 0)
+	nodesQueue = append(nodesQueue, rootCommitID)
+	currentGeneration := 1
+	// update commitNodes' generation in commitIDToNode "tree" using BFS algorithm.
+	// using a queue implementation
+	// adding a node to the queue only after all of its parents were visited in order to avoid redundant visits of nodes
+	for ; len(nodesQueue) > 0; currentGeneration++ {
+		var nodesIDsToIterateNextIteration []graveler.CommitID
+		for len(nodesQueue) > 0 {
+			currentCommitID := nodesQueue[0]
+			currentNode := commitIDToNode[currentCommitID]
+			commitIDToNode[currentCommitID] = CommitNode{currentNode.childCommitsIDs, 0, currentGeneration}
```

Comment:
This copies elements back into the node.  Can't we just say
```suggestion
			commitIDToNode[currentCommitID].generation = currentGeneration
			# (don't really need the next line, but if you like...)
			commitIDToNode[currentCommitID].notVisitedParents = 0
```
?

Similarly, I would rather modify just a single field on l. 513 than replace the entire `struct`.